lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -1,226 +1,272 @@
 # frozen_string_literal: true
 
+require 'forwardable'
+
 module KnapsackPro
   module Runners
     module Queue
       class RSpecRunner < BaseRunner
-        @@used_seed = nil
-
-        def self.run(args)
-          require 'rspec/core'
-          require_relative '../../formatters/time_tracker'
-          require_relative '../../formatters/time_tracker_fetcher'
-          require_relative '../../formatters/rspec_queue_summary_formatter'
-          require_relative '../../formatters/rspec_queue_profile_formatter_extension'
-
-          ENV['KNAPSACK_PRO_TEST_SUITE_TOKEN'] = KnapsackPro::Config::Env.test_suite_token_rspec
-          ENV['KNAPSACK_PRO_QUEUE_RECORDING_ENABLED'] = 'true'
-          ENV['KNAPSACK_PRO_QUEUE_ID'] = KnapsackPro::Config::EnvGenerator.set_queue_id
-
-          KnapsackPro::Config::Env.set_test_runner_adapter(adapter_class)
-          runner = new(adapter_class)
-
-          cli_args = (args || '').split
-          adapter_class.ensure_no_tag_option_when_rspec_split_by_test_examples_enabled!(cli_args)
-
-          # when format option is not defined by user then use progress formatter to show tests execution progress
-          cli_args += ['--format', 'progress'] unless adapter_class.has_format_option?(cli_args)
-
-          cli_args += [
-            # shows summary of all tests executed in Queue Mode at the very end
-            '--format', KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s,
-            '--format', KnapsackPro::Formatters::TimeTracker.to_s,
-            '--default-path', runner.test_dir,
-          ]
-
-          accumulator = {
-            status: :next,
-            runner: runner,
-            can_initialize_queue: true,
-            args: cli_args,
-            exitstatus: 0,
-            all_test_file_paths: [],
-          }
-          while accumulator[:status] == :next
-            handle_signal!
-            accumulator = run_tests(accumulator)
+        extend Forwardable
+
+        attr_reader :rspec_runner, :node_assigned_test_file_paths
+        attr_accessor :rspec_configuration_options
+
+        def_delegators :@rspec_runner, :world, :options, :configuration, :exit_code, :configure
+
+        def initialize(adapter_class)
+          super
+          @node_assigned_test_file_paths = []
+          @rspec_runner = nil # lazy assigned instance of ::RSpec::Core::Runner
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L85
+        #
+        # @return [Fixnum] exit status code.
+        #   0 if all specs passed,
+        #   or the configured failure exit code (1 by default) if specs failed.
+        def run(rspec_runner)
+          @rspec_runner = rspec_runner
+
+          rspec_runner.knapsack__setup
```

Comment:
The `rspec_runner` is an instance of `::RSpec::Core::Runner`. It has methods like `knapsack__setup` added by `KnapsackPro::Extensions::RSpecExtension.setup!`. 

I'd like to go in the following direction based on https://www.youtube.com/watch?v=yTkzNHF6rMs

I can inject the `rspec_runner` value with mocked `knapsack__*` methods in the test environment to test different logic paths for core functionalities of the knapsack_pro.

---

The `knapsack__*`  methods purpose is to use the RSpec internal methods that should not be used as public methods. RSpec has many public methods but they are [marked as internal API with comments](https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L129). We hide all the complexity of using the RSpec internal "private" methods inside of the  `knapsack__*`  methods.

I have isolation from RSpec internals and core knapsack logic. The core knapsack logic paths can be tested with unit tests and mocks.

The actual RSpec behavior can be tested when we run the knapsack_pro command on CI. For example, if something changes in the RSpec internals, then tests executed by the knapsack_pro command should fail on CI.

@3v0k4 Please let me know if you see something that is wrong and could be done better. Thank you.

Reply:
@ArturT 

I believe the spirit of that talk is different. You may want to watch the same talk in a [different version](https://www.destroyallsoftware.com/talks/boundaries); in particular from minute 29:00: Gary does not test the shell and pretty much does not use any test doubles because of that.

The functional core contains only pure functions and all the conditionals, so it's a breeze to test. The imperative shell contains only dependencies, and can be tested with "one" integration test because there's pretty much one path through it, and if it works once, it works all the time.

Now, in this case, the situation is a bit more messy, so I wouldn't try to be that perfect. But I'd see what happens with:
- Isolating pure functions and pushing all the conditionals inside them (or as many as possible), and unit testing them hard
- Keeping RSpec code in the extension in a shape that closely matches the original RSpec code, so that we don't lose the mapping
- Trying to have no conditionals outside of the shell, and test it at the integration level

The difficult part is finding the boundaries (and base the communication and values).

I hope this is helpful. Also, I'd be happy to pair program ro discuss some ideas as I don't have precise answers.

Reply:
TODO: second video to watch https://www.destroyallsoftware.com/talks/boundaries